misc/python/materialize/cli/mzcompose.py

```diff
@@ -177,142 +189,251 @@ def gen_shortcuts(repo: mzbuild.Repository) -> int:
+                # it as root.
+                (composition.path / "coverage").mkdir(exist_ok=True)
+            deps.acquire()
+
+        self.handle_composition(args, composition)
+
+    def handle_composition(
+        self, args: argparse.Namespace, composition: mzcompose.Composition
+    ) -> None:
+        announce("Delegating to Docker Compose")
+        proc = composition.run(
+            [*args.unknown_args, self.name, *args.unknown_subargs],
+            check=False,
         )
-        self.add_argument("--mz-coverage", action="store_true", default=None)
-        self.add_argument("-f", "--file")
-        self.add_argument("--project-directory")
-        self.add_argument("command", nargs="?")
-        self.add_argument("first_command_arg", nargs="?")
-        self.add_argument("remainder", nargs="*")
-        self.add_argument("--help", action="store_true", default=False)
+        sys.exit(proc.returncode)
+
+
+class RunCommand(DockerComposeCommand):
+    def __init__(self) -> None:
+        super().__init__("run", "run a one-off command")
+
+    def configure(self, parser: argparse.ArgumentParser) -> None:
+        super().configure(parser)
+        parser.add_argument("workflow", nargs="?")
+
+    def run(self, args: argparse.Namespace) -> None:
+        super().run(args)
+        if self.help:
+            print(
+                """As an mzcompose extension, run also supports running a workflow, as in:
+
+    $ ./mzcompose run WORKFLOW
+
+In this form, run does not accept any of the arguments listed above.
+
+To see the available workflows, run:
+
+    $ ./mzcompose list-services
+"""
+            )
+
+    def handle_composition(
+        self, args: argparse.Namespace, composition: mzcompose.Composition
+    ) -> None:
+        try:
+            workflow = composition.get_workflow(args.workflow, dict(os.environ))
+        except KeyError:
+            # Restart any dependencies whose definitions have changed. This is
+            # Docker Compose's default behavior for `up`, but not for `run`,
+            # which is a constant irritation that we paper over here. The trick,
+            # taken from Buildkite's Docker Compose plugin, is to run an `up`
+            # command that requests zero instances of the requested service.
+            composition.run(
+                ["up", "-d", "--scale", f"{args.workflow}=0", args.workflow]
+            )
+            super().handle_composition(args, composition)
+        else:
+            # The user has specified a workflow rather than a service. Run the
+            # workflow instead of Docker Compose.
+            unknown_args = [*args.unknown_args, *args.unknown_subargs]
+            if unknown_args:
+                raise errors.MzRuntimeError(
+                    f"cannot specify option {unknown_args[0]!r} when running a workflow"
+                )
+            workflow.run()
+
 
+BuildCommand = DockerComposeCommand("build", "build or rebuild services")
```

Comment:
these `DockerCompose` commands could add help in their `add_parser` call saying they just delegate

Reply:
👍🏽 I'll punt this one to a follow up too!

Reply:
sounds good!

Reply:
Actually knocked this one out now.